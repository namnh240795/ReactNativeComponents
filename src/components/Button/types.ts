import { Spacing } from '@src/atoms/theme/constants/common';
import {
  ButtonSize,
  ButtonType,
  ThemeColor,
} from '@src/atoms/theme/constants/types';
import { ButtonProps } from 'react-native';

export interface IButtonProps extends ButtonProps {
  size?: ButtonSize;
  type?: ButtonType;
  disabled?: boolean;
  margin?: Spacing;
  padding?: Spacing;
  fontSize?: number;
  children?: string;
  backgroundColor?: ThemeColor;
  textColor?: ThemeColor;
  borderRadius?: Spacing;
  width?: string | number | undefined;
  height?: string | number | undefined;
}
