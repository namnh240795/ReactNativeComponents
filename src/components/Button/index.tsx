/* eslint-disable react/react-in-jsx-scope */
import {
  BUTTON_TYPE,
  DEFAULT_BUTTON_STYLE,
  DISABLE_BUTTON_STYLE,
  getSizeButton,
} from '@src/atoms/theme/constants/button';
import { COLORS } from '@src/atoms/theme/constants/colors';
import { useMemo } from 'react';
import { stylesUtils } from '@src/atoms/theme/constants/common';
import { StyleSheet } from 'react-native';
import Box from '../Box';
import { Typography } from '../Typography';
import { IButtonProps } from './types';

const Button: React.FC<IButtonProps> = ({
  title,
  type,
  size,
  disabled,
  padding,
  margin,
  ...props
}) => {
  const typoStyle = useMemo(() => {
    if (disabled) {
      return COLORS.GOLBAT_60;
    } else {
      if (type === 'black05') {
        return COLORS.GOLBAT_80;
      }
      if (type === 'brand03') {
        return COLORS.BRAND_02;
      } else {
        return COLORS.GOLBAT_140;
      }
    }
  }, [disabled, type]);

  const buttonContainerStyle = useMemo(() => {
    let sizeValue, typeValue;
    let marginValues, paddingValues;
    switch (type) {
      case 'brand01':
        typeValue = BUTTON_TYPE.BRAND_01;
        break;
      case 'brand03':
        typeValue = BUTTON_TYPE.BRAND_03;
        break;
      case 'black05':
        typeValue = BUTTON_TYPE.BLACK_05;
        break;
      default:
        typeValue = DEFAULT_BUTTON_STYLE;
    }
    if (disabled) {
      typeValue = DISABLE_BUTTON_STYLE;
    }
    if (margin) {
      marginValues = stylesUtils.margin(stylesUtils.normalizeOptions(margin));
    }
    if (padding) {
      paddingValues = stylesUtils.padding(
        stylesUtils.normalizeOptions(padding),
      );
    }

    return StyleSheet.flatten([
      typeValue,
      sizeValue,
      marginValues,
      paddingValues,
      size && getSizeButton(size),
    ]);
  }, [size, type, margin, padding, disabled]);

  return (
    <Box style={buttonContainerStyle} pressable={!disabled} {...props}>
      <Typography color={typoStyle}>{title}</Typography>
    </Box>
  );
};

export default Button;
