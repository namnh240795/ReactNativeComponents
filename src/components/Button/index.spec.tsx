import React from 'react';
import renderer from 'react-test-renderer';
import Button from '.';

test('Render padding all button', () => {
  const tree = renderer.create(<Button title={''} padding={8} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render padding hv button', () => {
  const tree = renderer.create(<Button title={''} padding={[0, 2]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render padding width array button', () => {
  const tree = renderer
    .create(<Button title={''} padding={[0, 2, 0, 4]} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render margin all button', () => {
  const tree = renderer.create(<Button title={''} margin={8} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render margin hv button', () => {
  const tree = renderer.create(<Button title={''} margin={[0, 2]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render margin width array button', () => {
  const tree = renderer
    .create(<Button title={''} margin={[0, 2, 0, 4]} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render type button default', () => {
  const tree = renderer.create(<Button title={''} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render type button brand 01', () => {
  const tree = renderer.create(<Button title={''} type="brand01" />).toJSON();
  expect(tree).toMatchSnapshot();
});
test('Render type button brand 03', () => {
  const tree = renderer.create(<Button title={''} type="brand03" />).toJSON();
  expect(tree).toMatchSnapshot();
});
test('Render type button black 05', () => {
  const tree = renderer.create(<Button title={''} type="black05" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render size button medium', () => {
  const tree = renderer.create(<Button title={''} size="medium" />).toJSON();
  expect(tree).toMatchSnapshot();
});
test('Render size button large', () => {
  const tree = renderer.create(<Button title={''} size="large" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render button disable', () => {
  const tree = renderer.create(<Button title={''} disabled />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Button click event', () => {
  const onPressButton = jest.fn();
  const tree = renderer.create(
    <Button title={''} onPress={onPressButton}>
      Click here!
    </Button>,
  );
  expect(tree).toMatchSnapshot();
});
