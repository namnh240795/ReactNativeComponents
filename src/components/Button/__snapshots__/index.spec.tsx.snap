// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button click event 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "marginLeft": 16,
      "marginRight": 16,
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  >
    Click here!
  </Text>
</View>
`;

exports[`Render button disable 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E9ECEF",
      "borderRadius": 16,
      "display": "flex",
      "marginLeft": 16,
      "marginRight": 16,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
  title=""
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#ADB5BD",
      }
    }
  />
</View>
`;

exports[`Render margin all button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "marginBottom": 8,
      "marginLeft": 8,
      "marginRight": 8,
      "marginTop": 8,
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;

exports[`Render margin hv button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "marginBottom": 0,
      "marginLeft": 2,
      "marginRight": 2,
      "marginTop": 0,
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;

exports[`Render margin width array button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "marginBottom": 0,
      "marginLeft": 4,
      "marginRight": 2,
      "marginTop": 0,
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;

exports[`Render padding all button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "marginLeft": 16,
      "marginRight": 16,
      "opacity": 1,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingTop": 8,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;

exports[`Render padding hv button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "marginLeft": 16,
      "marginRight": 16,
      "opacity": 1,
      "paddingBottom": 0,
      "paddingLeft": 2,
      "paddingRight": 2,
      "paddingTop": 0,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;

exports[`Render padding width array button 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "marginLeft": 16,
      "marginRight": 16,
      "opacity": 1,
      "paddingBottom": 0,
      "paddingLeft": 4,
      "paddingRight": 2,
      "paddingTop": 0,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;

exports[`Render size button large 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "flex": 1,
      "marginLeft": 16,
      "marginRight": 16,
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;

exports[`Render size button medium 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "flex": 0.5,
      "marginLeft": 16,
      "marginRight": 16,
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
      "width": "50%",
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;

exports[`Render type button black 05 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F8F9FA",
      "borderRadius": 16,
      "display": "flex",
      "marginLeft": "auto",
      "marginRight": "auto",
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#6C757D",
      }
    }
  />
</View>
`;

exports[`Render type button brand 01 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "flex": 1,
      "marginLeft": "auto",
      "marginRight": "auto",
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;

exports[`Render type button brand 03 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFF3D9",
      "borderRadius": 16,
      "display": "flex",
      "marginLeft": "auto",
      "marginRight": "auto",
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#F88F22",
      }
    }
  />
</View>
`;

exports[`Render type button default 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FBB931",
      "borderRadius": 16,
      "display": "flex",
      "marginLeft": 16,
      "marginRight": 16,
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 24,
      "paddingRight": 24,
      "paddingTop": 12,
    }
  }
>
  <Text
    allowFontScaling={false}
    style={
      Object {
        "color": "#212529",
      }
    }
  />
</View>
`;
